File: src/main/java/extracells/container/ContainerHardMEDrive.java
package extracells.container;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.Container;
import net.minecraft.inventory.IInventory;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;
import extracells.container.slot.SlotRespective;
import extracells.tile.TileEntityHardMEDrive;

public class ContainerHardMEDrive extends Container
{

	IInventory tileentity;

	public ContainerHardMEDrive(InventoryPlayer inventory, TileEntityHardMEDrive tileentity)
	{
		this.tileentity = tileentity;
		for (int i = 0; i < 3; i++)
		{
			addSlotToContainer(new SlotRespective(tileentity, i, 80, 17 + i * 18)
			{
				public boolean isItemValid(ItemStack item)
				{
					return appeng.api.Util.getCellRegistry().isCellHandled(item);
				}
			});
		}
		bindPlayerInventory(inventory);
	}

	protected void bindPlayerInventory(InventoryPlayer inventoryPlayer)
	{
		for (int i = 0; i < 3; i++)
		{
			for (int j = 0; j < 9; j++)
			{
				addSlotToContainer(new Slot(inventoryPlayer, j + i * 9 + 9, 8 + j * 18, 84 + i * 18));
			}
		}

		for (int i = 0; i < 9; i++)
		{
			addSlotToContainer(new Slot(inventoryPlayer, i, 8 + i * 18, 142));
		}
	}

	@Override
	public boolean canInteractWith(EntityPlayer entityplayer)
	{
		return true;
	}

	@Override
	public ItemStack transferStackInSlot(EntityPlayer p, int i)
	{
		ItemStack itemstack = null;
		Slot slot = (Slot) inventorySlots.get(i);
		if (slot != null && slot.getHasStack())
		{
			ItemStack itemstack1 = slot.getStack();
			itemstack = itemstack1.copy();
			if (appeng.api.Util.getCellRegistry().isCellHandled(itemstack))
			{
				if (i < 3)
				{
					if (!mergeItemStack(itemstack1, 3, 38, false))
					{
						return null;
					}
				} else if (!mergeItemStack(itemstack1, 0, 3, false))
				{
					return null;
				}
				if (itemstack1.stackSize == 0)
				{
					slot.putStack(null);
				} else
				{
					slot.onSlotChanged();
				}
			}
		}
		return itemstack;
	}

	@Override
	public void onContainerClosed(EntityPlayer entityplayer)
	{
		super.onContainerClosed(entityplayer);
	}

	protected void retrySlotClick(int par1, int par2, boolean par3, EntityPlayer par4EntityPlayer)
	{
		// DONT DO ANYTHING YOU SHITTY METHOD!
	}
}

File: src/main/scala/extracells/container/ContainerHardMEDrive.scala
package extracells.container

import appeng.api.AEApi
import extracells.container.slot.SlotRespective
import extracells.tileentity.TileEntityHardMeDrive
import net.minecraft.entity.player.{EntityPlayer, InventoryPlayer}
import net.minecraft.inventory.{Container, Slot}
import net.minecraft.item.ItemStack


class ContainerHardMEDrive(inventory: InventoryPlayer, tile: TileEntityHardMeDrive) extends Container{


  for (i <- 0 to 2)
  {
    addSlotToContainer(new SlotRespective(tile.getInventory, i, 80, 17 + i * 18)
    {
      override def isItemValid(item: ItemStack) =  AEApi.instance.registries.cell().isCellHandled(item)
    });
  }

  bindPlayerInventory()

  protected def bindPlayerInventory(): Unit =
  {
    for (i <- 0 to 2)
    {
      for (j <- 0 to 8)
      {
        addSlotToContainer(new Slot(inventory, j + i * 9 + 9, 8 + j * 18, 84 + i * 18));
      }
    }

    for (i <- 0 to 8)
    {
      addSlotToContainer(new Slot(inventory, i, 8 + i * 18, 142));
    }
  }


  override def transferStackInSlot(p: EntityPlayer, i: Int): ItemStack = {
    var itemstack: ItemStack = null
    val slot = inventorySlots.get(i).asInstanceOf[Slot]
    if (slot != null && slot.getHasStack())
    {
      val itemstack1 = slot.getStack();
      itemstack = itemstack1.copy();
      if (AEApi.instance.registries.cell().isCellHandled(itemstack))
      {
        if (i < 3)
        {
          if (!mergeItemStack(itemstack1, 3, 38, false))
          {
            return null
          }
        } else if (!mergeItemStack(itemstack1, 0, 3, false))
        {
          return null
        }
        if (itemstack1.stackSize == 0)
        {
          slot.putStack(null)
        } else
        {
          slot.onSlotChanged()
        }
      }
    }
    itemstack
  }

  override def canInteractWith(p_75145_1_ : EntityPlayer): Boolean = {
    if (tile.hasWorldObj)
      tile.getWorldObj.getTileEntity(tile.xCoord, tile.yCoord, tile.zCoord) == this.tile
    else
      false
  }
}
